File: AlgorithmicToolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/2_maximum_value_of_the_loot.cpp
#include <iostream>
#include <algorithm>
#include <vector>

/*
*A thief finds much more loot than his bag can fit. Help him to find the most valuable
 combination of items assuming that any fraction of a loot item can be put into his bag.
*The goal of this code problem is to implement an algorithm for the fractional knapsack
 problem.
*/

using namespace std;

struct Item
{
    int a, b;
    double c = static_cast<double>(a) / b;
};

bool sortbylast(const Item& a, const Item& b) {
    return (a.c > b.c);
}

double get_optimal_value(int capacity, vector<int> weights, vector<int> values) {
    double value = 0.0;

    int size = weights.size();
    vector < Item > my_vec(size);
    vector <double> A(size, 0.0);


    for (int i = 0; i < size; ++i) {
        my_vec[i] = { values[i], weights[i]};
    }

    // sort according to value-weight ratio in reverse order.
    std::sort(my_vec.begin(), my_vec.end(), sortbylast);

    int W = capacity;
    for (int i = 0; i < weights.size(); ++i) {
        if (W == 0) return value;

        double a = min(my_vec[0].b, W);
        value += a * my_vec[0].c;
        my_vec[0].b -= a;

        W -= a;

        my_vec.erase(my_vec.begin());


    }

    return value;
}

int main() {
    int n;
    int capacity;
    std::cin >> n >> capacity;
    vector<int> values(n);
    vector<int> weights(n);
    for (int i = 0; i < n; i++) {
        std::cin >> values[i] >> weights[i];
    }

    double optimal_value = get_optimal_value(capacity, weights, values);

    std::cout.precision(10);
    std::cout << optimal_value << std::endl;


    return 0;
}

File: AlgorithmicToolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/2_maximum_value_of_the_loot.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


# Maximum value of the loot
# python3
from sys import stdin

"""
*A thief finds much more loot than his bag can fit. Help him to find the most valuable combination
 of items assuming that any fraction of a loot item can be put into his bag.
*The goal of this code problem is to implement an algorithm for the fractional knapsack problem.
"""

def maximum_loot_value(capacity, weights, prices):
    assert 0 <= capacity <= 2 * 10 ** 6
    assert len(weights) == len(prices)
    assert 1 <= len(weights) <= 10 ** 3
    assert all(0 < w <= 2 * 10 ** 6 for w in weights)
    assert all(0 <= p <= 2 * 10 ** 6 for p in prices)
    
    A = [0]*len(weights)
    V = 0
    
    W = capacity
    for i in range(len(weights)):
        if W == 0:
            return V
        prices, weights = zip(*sorted(zip(prices, weights), key=lambda t: t[0]/t[1], reverse=True))
        
        prices = list(prices)
        weights = list(weights)
        
        a = min(weights[0], W)
        V += a*(prices[0]/weights[0])
        weights[0] -= a
        A[i] += a
        W -= a
        del prices[0]
        del weights[0]
    return V
        
        
if __name__ == "__main__":
    data = list(map(int, stdin.read().split()))
    n, input_capacity = data[0:2]
    input_prices = data[2:(2 * n + 2):2]
    input_weights = data[3:(2 * n + 2):2]
    opt_value = maximum_loot_value(input_capacity, input_weights, input_prices)
    print("{:.10f}".format(opt_value))

